### PR TITLE
Move memory info into IAllocator

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
  - Update the IAllocator setup to move the OrtMemoryInfo to the base class instead of requiring derived classes to have that as a member and override a virtual method to return it.
  - Cleanup CreateAllocator setup to take an argument as to whether to wrap the device allocator in an arena allocator. 
    - Whether a device allocator instance gets wrapped by the arena allocator is a choice made at a higher level than the device allocator, so getting the value from IDeviceAllocator::AllowsArena() caused a confusing setup. 
  - Add ability to disable using an arena for CUDA allocations for consistency with what's allowed for the CPU allocations
    - update relevant EPs to check this flag
    - add command line param to disable arena to test exes
  - Minor cleanups in the various EPs to adjust to the change to IAllocator and CreateAllocator, and to use the create_area flag consistently when available.
  - Move DummyArena to test code. It shouldn't be needed by any production code anymore.

**Motivation and Context**
Initial motivation was adding the ability to disable the arena for CUDA allocations for consistency. In making that change the confusing setup around how/when an arena allocator wrapped the device allocator led to the IAllocator changes.

### CODE REVIEW

File: include/onnxruntime/core/framework/allocator.h [hunk 1]
```diff
@@ -258,50 +265,30 @@ bool IAllocator::CalcMemSizeForArrayWithAlignment(size_t nmemb, size_t size, siz
...
-    memory_info_ = onnxruntime::make_unique<OrtMemoryInfo>(CPU, OrtAllocatorType::OrtDeviceAllocator);
-  }
+  CPUAllocator() : IDeviceAllocator(OrtMemoryInfo(CPU, OrtAllocatorType::OrtDeviceAllocator)) {}
 
   void* Alloc(size_t size) override;
   void Free(void* p) override;
-  const OrtMemoryInfo& Info() const override;
-
- private:
-  std::unique_ptr<OrtMemoryInfo> memory_info_;
```
Comment (pranavsharma) [inline include/onnxruntime/core/framework/allocator.h]: Good call on removing the heap allocation as well.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 2]
```diff
@@ -296,12 +296,18 @@ struct OrtApi {
...
+  // Enable a memory arena for CPU allocations.
   // Arena may pre-allocate memory for future usage.
   // set this option to false if you don't want it.
   OrtStatus*(ORT_API_CALL* EnableCpuMemArena)(_Inout_ OrtSessionOptions* options)NO_EXCEPTION;
   OrtStatus*(ORT_API_CALL* DisableCpuMemArena)(_Inout_ OrtSessionOptions* options)NO_EXCEPTION;
 
+  // Enable a memory arena for CUDA allocations.
+  // Arena may pre-allocate memory for future usage.
+  // set this option to false if you don't want it.
+  OrtStatus*(ORT_API_CALL* EnableCudaMemArena)(_Inout_ OrtSessionOptions* options)NO_EXCEPTION;
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h]: Does this mean we need to add more such calls for every EP?

File: onnxruntime/core/framework/allocatormgr.cc [hunk 3]
```diff
@@ -10,20 +10,20 @@
...
-  using TArenaAllocator = BFCArena;
+using TArenaAllocator = BFCArena;
 #endif
 
-using namespace ::onnxruntime::common;
-
-AllocatorPtr CreateAllocator(DeviceAllocatorRegistrationInfo info, int device_id) {
+AllocatorPtr CreateAllocator(DeviceAllocatorRegistrationInfo info, int device_id, bool use_arena) {
   auto device_allocator = std::unique_ptr<IDeviceAllocator>(info.factory(device_id));
-  if (device_allocator->AllowsArena()) {
```
Comment (pranavsharma) [inline onnxruntime/core/framework/allocatormgr.cc]: Can we assume that all devices are amenable to arena type allocation? If not, the use_arena when true could be misleading.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 2 above)
Comment (skottmckay) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: I don't think so. AFAIK it's only the CPU and CUDA EPs that use BFCArena currently. If additional EPs start to use it we should have a more generic way of enabling/disabling the arena for each (I think we need a way to specify these sort of finer granularity things without adding a new function or variable to the API each time).

File: onnxruntime/core/framework/allocatormgr.cc (hunk 3 above)
Comment (skottmckay) [reply onnxruntime/core/framework/allocatormgr.cc]: I believe so. The arena is essentially a wrapper class on top of the device allocator, and both implement the IAllocator interface. The device allocator doesn't and shouldn't need to care about whether it's getting Alloc/Free calls from an arena layer or not. The EP should be the one to decide whether it wants the arena wrapper or not, so I think providing use_arena as a parameter is clear.

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 4]
```diff
@@ -109,10 +109,10 @@ std::string nuphar_settings;
 #endif
 
 namespace onnxruntime {
-std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_CPU(int use_arena);
-std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_CUDA(int device_id);
-std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_Tensorrt(int device_id);
-std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_Dnnl(int use_arena);
+std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_CPU(bool use_arena);
+std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_CUDA(int device_id, bool use_cuda_arena, bool use_cpu_arena);
```
Comment (pranavsharma) [inline onnxruntime/python/onnxruntime_pybind_state.cc]: nit: easy to mess up (reverse) the values of these (use_cuda_arena and use_cpu_arena) at the call site.

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 5]
```diff
@@ -402,12 +413,12 @@ void addGlobalMethods(py::module& m) {
                 /*default_max_vlog_level*/ -1);
 
         std::vector<std::shared_ptr<onnxruntime::IExecutionProviderFactory>> factories = {
-            onnxruntime::CreateExecutionProviderFactory_CPU(0),
+            onnxruntime::CreateExecutionProviderFactory_CPU(true),
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/python/onnxruntime_pybind_state.cc]: Not sure it matters  - was just trying to be a little more consistent (CUDA and DNNL were using arenas) here. If this is just to dump schemas none of them should need an arena so I will try setting them all to not do so.

---
In reply to: [372029154](https://github.com/microsoft/onnxruntime/pull/2850#discussion_r372029154) [](ancestors = 372029154)

File: onnxruntime/python/onnxruntime_pybind_state.cc (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/python/onnxruntime_pybind_state.cc]: Do you want them combined into an enum? At least it's slightly clearer than the BrainSlice one. :)

---
In reply to: [372065123](https://github.com/microsoft/onnxruntime/pull/2850#discussion_r372065123) [](ancestors = 372065123)

File: onnxruntime/python/onnxruntime_pybind_state.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnxruntime/python/onnxruntime_pybind_state.cc]: Ok, yeah this looks like inside the dump schema condition. Shouldn't matter. Just seeing 0 replaced with true made me wonder.

File: onnxruntime/python/onnxruntime_pybind_state.cc (hunk 4 above)
Comment (pranavsharma) [reply onnxruntime/python/onnxruntime_pybind_state.cc]: Up to you. Doesn't have to be this PR. The BrainSlice interface is poorly designed and is not a good reference. Just for reference: https://www.aristeia.com/Papers/IEEE_Software_JulAug_2004_revised.htm

**Comment by pranavsharma** [comment]
> Not finished fully reviewing. Will look into it more early tomorrow.

done reviewing.

File: include/onnxruntime/core/framework/allocator.h [hunk 6]
```diff
@@ -155,13 +156,14 @@ using IAllocatorUniquePtr = std::unique_ptr<T, std::function<void(T*)>>;
 
 class IAllocator {
  public:
+  IAllocator(const OrtMemoryInfo& info) : memory_info_(info) {}
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply include/onnxruntime/core/framework/allocator.h:159]: How does this change with a copy vs std::move? OrtMemoryInfo doesn't own or free the const char* value whether there's a copy or a move. 

If we want it to, OrtMemoryInfo::name needs to be a std::string.

[the remark this answers is not in the document]
Comment (pranavsharma) [reply include/onnxruntime/core/framework/allocator.h:159]: Yes, free is a problem too. There's no documentation about the ownership semantics. If this class is designed to have value semantics, we should preferably use std::string.

**Pull request merged**